src/minimega/kvm.go

```diff
@@ -520,54 +490,48 @@ func (vm *KvmVM) launch(ack chan int) (err error) {
 
 	vm.CheckAffinity()
 
+	// Channel to signal when the process has exited
+	var waitChan = make(chan bool)
+
+	// Create goroutine to wait for process to exit
 	go func() {
+		defer close(waitChan)
 		err := cmd.Wait()
-		vm.setState(VM_QUIT)
-		if err != nil {
-			if err.Error() != "signal: killed" { // because we killed it
-				log.Error("kill qemu: %v %v", err, sErr.String())
-				vm.setState(VM_ERROR)
-			}
-		}
-		waitChan <- vm.ID
-	}()
 
-	// we can't just return on error at this point because we'll leave dangling goroutines, we have to clean up on failure
-	sendKillAck := false
+		vm.Lock()
+		defer vm.Unlock()
 
-	// connect to qmp
-	connected := false
-	for count := 0; count < QMP_CONNECT_RETRY; count++ {
-		vm.q, err = qmp.Dial(vm.qmpPath())
-		if err == nil {
-			connected = true
-			break
+		// Check if the process quit for some reason other than being killed
+		if err != nil && err.Error() != "signal: killed" {
+			log.Error("kill qemu: %v %v", err, sErr.String())
+			vm.setError(err)
+		} else if vm.State != VM_ERROR {
+			// Set to QUIT unless we've already been put into the error state
+			vm.setState(VM_QUIT)
 		}
-		delay := QMP_CONNECT_DELAY * time.Millisecond
-		log.Info("qmp dial to %v : %v, redialing in %v", vm.ID, err, delay)
-		time.Sleep(delay)
-	}
+	}()
 
-	if !connected {
+	if err := vm.connectQMP(); err != nil {
+		// Failed to connect to qmp so clean up the process
 		cmd.Process.Kill()
+
 		log.Errorln(err)
-		return fmt.Errorf("vm %v failed to connect to qmp: %v", vm.ID, err)
+		vm.setError(err)
+		return err
 	}
 
-	log.Debug("qmp dial to %v successful", vm.ID)
-
 	go vm.asyncLogger()
 
-	ack <- vm.ID
-
 	// connect cc
 	ccPath := filepath.Join(vm.instancePath, "cc")
-	err = ccNode.DialSerial(ccPath)
-	if err != nil {
+	if err := ccNode.DialSerial(ccPath); err != nil {
+		// TODO: Should we kill the VM?
```

Comment:
No, we should simply warn on the error. There are scenarios where people will use the qemu override to disable the cc virtio port for various reasons. Not common enough to justify making the cc port a tunable, but just the same it shouldn't prevent folks from still running a vm.